README.md
# Sequel Datastore ORM

Inspired by Laravel Eloquent and Google Cloud Datastore

## Database Support

- [x] MySQL (version 5.7 and above)
- [x] Postgres (version 9.4 and above)

## Installation

```bash
  // dependency
  $ go get -u github.com/go-sql-driver/mysql // Mysql
  $ go get -u github.com/lib/pq // Postgres
  $ go get -u cloud.google.com/go/datastore
  $ go get -u github.com/si3nloong/goloquent
```

- **Import the library**

```go
  import "github.com/si3nloong/goloquent"
```

## Go Support

At least 1.10, because we use `strings.Builder` and `sync.Map`.

## Quick Start

### Connect to database

```go
    import "github.com/si3nloong/goloquent/db"

    conn, err := db.Open("mysql", db.Config{
        Username: "root",
        Password: "",
        Host: "localhost",
        Port: "3306",
        Database: "test",
        Logger: func(stmt *goloquent.Stmt) {
            log.Println(stmt.TimeElapse()) // elapse time in time.Duration
            log.Println(stmt.String()) // Sql string without any ?
            log.Println(stmt.Raw()) // Sql prepare statement
            log.Println(stmt.Arguments()) // Sql prepare statement's arguments
            log.Println(fmt.Sprintf("[%.3fms] %s", stmt.TimeElapse().Seconds()*1000, stmt.String()))
        },
    })
    defer conn.Close()
    if err != nil {
        panic("Connection error: ", err)
    }
```

#### User Table

```go
// Address :
type Address struct {
	Line1    string
	Line2    string
	Country  string
	PostCode uint
	Region   struct {
		TimeZone    time.Time
		Keys        []*datastore.Key   `goloquent:"keys"`
		CountryCode string             `goloquent:"regionCode"`
		Geolocation datastore.GeoPoint `goloquent:"geo"`
	} `goloquent:"region"`
}

// User : User kind parent is Merchant
type User struct {
    Key             *datastore.Key `goloquent:"__key__"` // load table key
    Name            string
    Nicknames       []string
    CountryCode     string
    PhoneNumber     string
    Email           string
    BirthDate       goloquent.Date
    Address         Address
    Age             uint8
    ExtraInfo       json.RawMessage
    CreatedDateTime time.Time
    UpdatedDateTime time.Time
}

// Load : load func will execute after load
func (x *User) Load() error {
	return nil
}

// Save : save func will execute before save
func (x *User) Save() (error) {
	return nil
}
```

### Helper

```go
    // StringPrimaryKey is to get key value in string form
    key := datastore.NameKey("Merchant", "mjfFgYnxBS", nil)
    fmt.Println(goloquent.StringKey(key)) // "mjfFgYnxBS"

    key = datastore.IDKey("User", int64(2305297334603281546), nil)
    fmt.Println(goloquent.StringKey(key)) // "2305297334603281546"

    key = datastore.NameKey("User", int64(2305297334603281546), datastore.NameKey("Merchant", "mjfFgYnxBS", nil))
    fmt.Println(goloquent.StringifyKey(key)) // Merchant,'mjfFgYnxBS'/User,2305297334603281546
```

### Table

```go
    import "github.com/si3nloong/goloquent/db"

    // Check table exists
    db.Table("User").Exists() // true

    // Drop table if exists
    if err := db.Table("User").DropIfExists(); err != nil {
        log.Fatal(err)
    }

    // Add index
    if err := db.Table("User").AddIndex("Name", "Email"); err != nil {
        log.Fatal(err)
    }

    // Add unique index
    if err := db.Table("User").AddUniqueIndex("Email"); err != nil {
        log.Fatal(err)
    }
```

### Create Record

```go
    import "github.com/si3nloong/goloquent/db"

    // Example
    user := new(User)
    user.Name = "Hello World"
    user.Age = 18

    // OR
    var user *User
    user.Name = "Hello World"
    user.Age = 18

    // Create without parent key
    if err := db.Create(user, nil); err != nil {
        log.Println(err) // fail to create record
    }

    // Create with parent key
    parentKey := datastore.NameKey("Parent", "value", nil)
    if err := db.Create(user, parentKey); err != nil {
        log.Println(err) // fail to create record
    }

    // Create without key, goloquent will auto generate primary key
    if err := db.Create(user); err != nil {
        log.Println(err) // fail to create record
    }

    // Create with self generate key
    user.Key = datastore.NameKey("User", "uniqueID", nil)
    if err := db.Create(user); err != nil {
        log.Println(err) // fail to create record
    }
```

### Upsert Record

```go
    import "github.com/si3nloong/goloquent/db"

    // Example
    user := new(User)
    user.Name = "Hello World"
    user.Age = 18

    // Update if key exists, else create the user record
    parentKey := datastore.NameKey("Parent", "value", nil)
    if err := db.Upsert(user, parentKey); err != nil {
        log.Println(err) // fail
    }

    // Upsert with self generate key
    user := new(User)
    user.Key = datastore.NameKey("User", "uniqueID", nil)
    user.Name = "Hello World"
    user.Age = 18
    if err := db.Upsert(user); err != nil {
        log.Println(err) // fail
    }
```

### Retrieve Record

- **Get Single Record using Primary Key**

```go
    // Example
    primaryKey := datastore.IDKey("User", int64(2305297334603281546), nil)
    user := new(User)
    if err := db.Find(primaryKey, user); err != nil {
        log.Println(err) // error while retrieving record
    }

    if err := db.Where("Status", "=", "ACTIVE").
        Find(primaryKey, user); err != goloquent.ErrNoSuchEntity {
        // if no record found using primary key, error `ErrNoSuchEntity` will throw instead
        log.Println(err) // error while retrieving record
    }
```

- **Get Single Record**

```go
    import "github.com/si3nloong/goloquent/db"
    // Example 1
    user := new(User)
    if err := db.First(user); err != nil {
        log.Println(err) // error while retrieving record
    }

    if user.Key != nil { // if have record
        fmt.Println("Have record")
    } else { // no record
        fmt.Println("Doesnt't have record")
    }

    // Example 2
    user := new(User)
    if err := db.Where("Email", "=", "admin@hotmail.com").
        First(user); err != nil {
        log.Println(err) // error while retrieving record
    }

    // Example 3
    age := 22
    parentKey := datastore.IDKey("Parent", 1093, nil)
    user := new(User)
    if err := db.Ancestor(parentKey).
        WhereEqual("Age", &age).
        Order("-CreatedDateTime").
        First(user); err != nil {
        log.Println(err) // error while retrieving record
    }
```

- **Get Multiple Record**

```go
    import "github.com/si3nloong/goloquent/db"
    // Example 1
    users := new([]User)
    if err := db.Limit(10).Get(users); err != nil {
        log.Println(err) // error while retrieving record
    }

    // Example 2
    users := new([]*User)
    if err := db.WhereEqual("Name", "Hello World").
        Get(users); err != nil {
        log.Println(err) // error while retrieving record
    }

    // Example 3
    users := new([]User)
    if err := db.Ancestor(parentKey).
        WhereEqual("Name", "myz").
        WhereEqual("Age", 22).
        Get(users); err != nil {
        log.Println(err) // error while retrieving record
    }
```

- **Get Record with Ordering**

```go
    import "github.com/si3nloong/goloquent/db"
    // Ascending order
    users := new([]*User)
    if err := db.Order("CreatedDateTime").
        Get(users); err != nil {
        log.Println(err) // error while retrieving record
    }

    // Descending order
    if err := db.Table("User").
        Order("-CreatedDateTime").
        Get(users); err != nil {
        log.Println(err) // error while retrieving record
    }
```

- **Pagination Record**

```go
    import "github.com/si3nloong/goloquent/db"

    p := goloquent.Pagination{
        Limit:  10,
        Cursor: "", // pass the cursor that generate by the query so that it will display the next record
    }

    // Example
    users := new([]*User)
    if err := db.Ancestor(parentKey).
        Order("-CreatedDateTime").
        Paginate(&p, users); err != nil {
        log.Println(err) // error while retrieving record
    }

    // ***************** OR ********************
    p := &goloquent.Pagination{
        Limit:  10, // number of records in each page
        Cursor: "EhQKCE1lcmNoYW50EK3bueKni5eNIxIWCgxMb3lhbHR5UG9pbnQaBkZrUUc4eA", // pass the cursor to get next record set
    }

    users := new([]*User)
    if err := db.Ancestor(parentKey).
        Order("-CreatedDateTime").
        Paginate(p, users); err != nil {
        log.Println(err) // error while retrieving record
    }

    log.Println(p.NextCursor()) // next page cursor
    log.Println(p.Count()) // record count
```

### Save Record

```go
    import "github.com/si3nloong/goloquent/db"
    // Example
    if err := db.Save(user); err != nil {
        log.Println(err) // fail to delete record
    }
```

### Delete Record

- **Delete using Primary Key**

```go
    import "github.com/si3nloong/goloquent/db"
    // Example
    if err := db.Delete(user); err != nil {
        log.Println(err) // fail to delete record
    }
```

- **Delete using Where statement**

```go
    // Delete user table record which account type not equal to "PREMIUM" or "MONTLY"
    if err := db.Table("User").
        WhereNotIn("AccountType", []string{
            "PREMIUM", "MONTLY",
        }).Flush(); err != nil {
        log.Println(err) // fail to delete record
    }
```

### Transaction

```go
    // Example
    if err := db.RunInTransaction(func(txn *goloquent.DB) error {
        user := new(User)
        if err := txn.Create(user, nil); err != nil {
            return err // return any err to rollback the transaction
        }
        return nil // return nil to commit the transaction
    }); err != nil {
        log.Println(err)
    }
```

- **Table Locking (only effective inside RunInTransaction)**

```go
    // Example
    merchantKey := datastore.IDKey("Merchant", "mjfFgYnxBS", nil)
    userKey := datastore.IDKey("User", int64(4645436182170916864), nil)
    if err := db.RunInTransaction(func(txn *goloquent.DB) error {
        user := new(User)

        if err := txn.NewQuery().
            WLock(). // Lock record for update
            Find(userKey, user); err != nil {
            return err
        }

        merchant := new(Merchant)
        if err := txn.NewQuery().
            RLock(). // Lock record for read
            Find(merchantKey, merchant); err != nil {
            return err
        }

        user.Age = 30
        if err := txn.Save(user); err != nil {
            return err // return any err to rollback the transaction
        }

        return nil // return nil to commit the transaction
    }); err != nil {
        log.Println(err)
    }
```

- **Database Migration**

```go
    import "github.com/si3nloong/goloquent/db"
    // Example
    user := new(User)
    if err := db.Migrate(
        new(user),
        Merchant{},
        &Log{},
    ); err != nil {
        log.Println(err)
    }
```

- **Filter Query**

```go
    import "github.com/si3nloong/goloquent/db"
    // Update single record
    user := new(User)
    if err := db.NewQuery().
        WhereIn("Status", []interface{}{"active", "pending"}).
        First(user); err != nil {
        log.Println(err) // error while retrieving record or record not found
    }

    // Get record with like
    if err := db.NewQuery().
        WhereLike("Name", "%name%").
        First(user); err != nil {
        log.Println(err) // error while retrieving record or record not found
    }
```

- **Update Query**

```go
    import "github.com/si3nloong/goloquent/db"

CHANGELOG.md
# Bug / Issue

- Set primary key on insert
- Primary key at insertion is wrong, always create new key
- Load relative / Eager loading
- Not checking present of `__key__`
- Not checking `SoftDelete`
- (2018-06-03) Fix empty table name when using `Flush` func.
- (2018-06-10) Fix panic occur on func `StringKey` when input parameter `*datastore.Key` is `nil`
- (2018-06-18) Primary key should omitted in operation `Upsert`
- (2018-06-18) Fix logger `String` func is output unexpected string when using `postgres` driver
- (2018-06-19) Fix flatten struct bug, flatten column using root data type instead of the subsequent data type
- (2018-06-19) Fix primary key bug when using `WHERE $Key IN (?)`, key is not convert to primary key format
- (2018-06-21) Fix alter table character set and collation bug, change from `ALTER TABLE xxx CONVERT TO CHARACTER SET utf8` to `ALTER TABLE xxx CHARACTER SET utf8`
- (2018-06-21) Fix mysql panic even is 5.7 or above `eg: GAE return 5.7.14-google-log instead 5.7.14` will mismatch in the string comparison
- (2018-06-22) Fix `Paginate` bug, model slice is appending instead of get replace
- (2018-06-25) Fix struct property sequence bug
- (2018-06-27) Fix struct codec, func `Select` and func `DistinctOn`
- (2018-06-28) Fix `postgres` update with limit clause bug. Only mysql support `UPDATE xxx SET xxx LIMIT 10`. Postgres instead will use `UPDATE xxx SET xxx WHERE key IN (SELECT xxx FROM xxx LIMIT 10)`.
- (2018-06-28) Fix `Paginate` bug, invalid cursor signature due to `qson` package didn't sort the filter fields
- (2018-07-02) Fix `panic: reflect: Field index out of range` on embeded struct, code paths is invalid
- (2018-07-02) Fix entity doesn't execute `Save` func even it implement `Saver` interface when it's not a pointer struct (eg: []Struct)
- (2018-07-05) Fix `postgres` `GetColumns` bug, it return empty array even database have records
- (2018-07-11) Fix `Update` func bug. It doesn't marshal the map[string]interface nor []interface{} to string after normalization
- (2018-07-13) Fix func `Unmarshal` of data type `Date`. It suppose using `YYYY-MM-DD` format.
- (2018-07-17) Fix panic when value of `WhereIn` or `WhereNotIn` contains `nil` value.
- (2018-07-17) Fix `First` func bug. Entity value doesn't override if the result is empty.
- (2018-07-18) Fix panic when `Where` value is pointer of `int`, `int8`, `int16`, `int32`, `uint`, `uint8`, `uint16`, `uint32`, `float32`.
- (2018-07-23) Fix `DB` connection bug when passing empty port and postgres unable to establish connection thru unix socket.
- (2018-08-16) Fix panic when `nil` value passing in with func `Create` or `Upsert`, it should return error instead.
- (2018-08-17) Fix invalid sql statement on `Paginate()` when using next `Cursor` from `NextCursor()`.
- (2018-08-23) Fix unicode string cannot save to `mysql`.
- (2018-09-06) Fix incorrect mysql schema for signed and unsigned integer data type.
  <!-- - (2018-09-10) Fix `Upsert` bug. Primary key should omitted. -->
- (2018-09-10) Fix `postgres` schema bug. Schema for unsigned integer should be greater and equal to zero instead of greater than zero. `CHECK (value >= 0)`.
- (2018-09-13) Fix `newPrimaryKey` logic error. ID key with 0 shouldn't nested again.

# Breaking Changes

- Drop `Count` api.
- Drop `Union` api.
- Drop single `Update` api.
- Drop `SetDebug(boolean)` api.
- Drop `datastore` driver support.
- `Delete` api using entity model instead of `*datastore.Key`.
- Rename params in function `RunInTransaction` from `*goloquent.Connection` to `*goloquent.DB`.
- Rename api `LockForUpdate` to `WLock`.
- Rename api `LockForShared` to `RLock`.
- Change single entity api `Update` to `Save`.
- Change `Loader` interface `Load([]datastore.Property) error` to `Load() error`
- Change `Saver` interface `Save() ([]datastore.Property,error)` to `Save() error`
- Change second parameter **parentKey** `*datastore.Key` to optional on function `Create` nor `Upsert`
- (2018-06-16) No longer support mysql 5.6 and below (at least 5.7).
- (2018-06-19) Table is now by default using `utf8mb4` encoding.
- (2018-06-20) Replace `Next` func in `Pagination` struct with `NextCursor`.
- (2018-06-21) Support extra option `datatype`, `charset`, `collate` on struct property, but it only limited to datatype of `string`.
- (2018-06-21) Allow `*` on api `Select`.
- (2018-06-24) Replace offset pagination with cursor pagination
- (2018-07-05) Rename api `WhereNe` to `WhereNotEqual`.
- (2018-07-08) Rename api `WhereEq` to `WhereEqual`.
- (2018-07-08) Replace return parameter `Query` to `Table` on api `Table` of `goloquent.DB`.
- (2018-07-17) Expose operator to public.
- (2018-08-24) Change `AnyOfAncestor` api parameter data type `[]*datastore.Key` to `...*datastore.Key`.
- (2018-08-24) Change `Truncate` api to support multiple entity.
- (2018-08-24) Increase maximum limit of `Pagination` api to `10000`.
- (2018-09-02) Change api `WhereIn` and `WhereNotIn` value param from `[]interface{}` to `interface{}`.
- (2018-09-10) `Date` no longer convert to UTC before format.
- (2018-09-10) `Date` will have default value `"0001-01-01"` if it's not pointer.
- (2018-09-13) Name key will escape using `url.PathEscape` to avoid misinterpret when it contains symbol characters.

# New Features

- Introduce `Select` api.
- Introduce `DistinctOn` api.
- Introduce `Lock` api.
- Introduce `Truncate` api.
- Introduce `Flush` api.
- Introduce package `db`.
- Introduce package `qson`. (Query JSON).
- Replace statement debug using `LogHandler`.
- Support unsigned integer, uint, uint8, uint16, uint32, uint64
- Support any pointer of base data type and struct
- (2018-06-14) Support **Postgres** driver.
- (2018-06-18) Introduce `Scan` api.
- (2018-06-22) Introduce hard delete api `Destroy`.

dialect_postgres.go
package goloquent

import (
	"bytes"
	"database/sql"
	"encoding/json"
	"fmt"
	"log"
	"reflect"
	"strings"
	"time"
)

type postgres struct {
	sequel
}

var _ Dialect = new(postgres)

func init() {
	RegisterDialect("postgres", new(postgres))
}

func (p postgres) escapeSingleQuote(n string) string {
	return strings.Replace(n, `'`, `\'`, -1)
}

// Open :
func (p *postgres) Open(conf Config) (*sql.DB, error) {
	buf := new(strings.Builder)
	buf.WriteString(fmt.Sprintf("user='%s' ", p.escapeSingleQuote(conf.Username)))
	buf.WriteString(fmt.Sprintf("password='%s' ", p.escapeSingleQuote(conf.Password)))
	if conf.UnixSocket != "" {
		buf.WriteString(fmt.Sprintf("host=/%s ", strings.Trim(conf.UnixSocket, `/`)))
	} else {
		host, port := "localhost", "5432"
		if conf.Host != "" {
			host = conf.Host
		}
		if conf.Port != "" {
			port = conf.Port
		}
		buf.WriteString(fmt.Sprintf("host=%s port=%s ", host, port))
	}
	buf.WriteString(fmt.Sprintf("dbname='%s' ", p.escapeSingleQuote(conf.Database)))
	buf.WriteString("sslmode=disable")
	log.Println("Connection String :", buf.String())
	client, err := sql.Open("postgres", buf.String())
	if err != nil {
		return nil, err
	}
	return client, nil
}

// GetTable :
func (p postgres) GetTable(name string) string {
	return p.Quote(name)
}

// CurrentDB :
func (p *postgres) CurrentDB() (name string) {
	if p.dbName != "" {
		name = p.dbName
		return
	}

	p.db.QueryRow("SELECT current_database();").Scan(&name)
	p.dbName = name
	return
}

func (p postgres) Quote(n string) string {
	return fmt.Sprintf(`"%s"`, n)
}

func (p postgres) Bind(i uint) string {
	return fmt.Sprintf("$%d", i)
}

func (p postgres) SplitJSON(name string) string {
	paths := strings.SplitN(name, ">", 2)
	if len(paths) <= 1 {
		return p.Quote(paths[0])
	}
	vv := strings.Split(strings.TrimSpace(paths[1]), `.`)
	return fmt.Sprintf(`%s->%s`,
		p.Quote(strings.TrimSpace(paths[0])),
		`'`+strings.Join(vv, p.Value(`->`))+`'`)
}

func (p postgres) JSONMarshal(v interface{}) (b json.RawMessage) {
	switch vi := v.(type) {
	case json.RawMessage:
		return vi
	case nil:
		b = json.RawMessage("null")
	case string:
		b = json.RawMessage(fmt.Sprintf("%q", vi))
	default:
		b = json.RawMessage(fmt.Sprintf("%v", vi))
	}
	return
}

func (p postgres) FilterJSON(f Filter) (string, []interface{}, error) {
	vv, err := f.Interface()
	if err != nil {
		return "", nil, err
	}
	if vv == nil {
		vv = json.RawMessage("null")
	}
	name := p.SplitJSON(f.Field())
	buf, args := new(bytes.Buffer), make([]interface{}, 0)
	switch f.operator {
	case Equal:
		buf.WriteString(fmt.Sprintf("(%s) = %s", name, variable))
	case NotEqual:
		buf.WriteString(fmt.Sprintf("(%s) <> %s", name, variable))
	case GreaterThan:
		buf.WriteString(fmt.Sprintf("(%s) > %s", name, variable))
	case GreaterEqual:
		buf.WriteString(fmt.Sprintf("(%s) >= %s", name, variable))
	case In:
		x, isOk := vv.([]interface{})
		if !isOk {
			x = append(x, vv)
		}
		if len(x) <= 0 {
			return "", nil, fmt.Errorf(`value for "In" operator cannot be empty`)
		}
		buf.WriteString("(")
		for i := 0; i < len(x); i++ {
			buf.WriteString(fmt.Sprintf("(%s = %s) OR ", name, variable))
			args = append(args, p.JSONMarshal(x[i]))
		}
		buf.Truncate(buf.Len() - 4)
		buf.WriteString(")")
		return buf.String(), args, nil
	case NotIn:
		x, isOk := vv.([]interface{})
		if !isOk {
			x = append(x, vv)
		}
		if len(x) <= 0 {
			return "", nil, fmt.Errorf(`value for "In" operator cannot be empty`)
		}
		buf.WriteString("(")
		for i := 0; i < len(x); i++ {
			buf.WriteString(fmt.Sprintf("(%s <> %s) AND ", name, variable))
			args = append(args, p.JSONMarshal(x[i]))
		}
		buf.Truncate(buf.Len() - 5)
		buf.WriteString(")")
		return buf.String(), args, nil
	case ContainAny:
		x, isOk := vv.([]interface{})
		if !isOk {
			x = append(x, vv)
		}
		if len(x) <= 0 {
			return "", nil, fmt.Errorf(`value for "In" operator cannot be empty`)
		}
		buf.WriteString(fmt.Sprintf("%s ?| array[", name))
		for i := 0; i < len(x); i++ {
			buf.WriteString(variable + ",")
			args = append(args, x[i])
		}
		buf.Truncate(buf.Len() - 1)
		buf.WriteString("]")
		return buf.String(), args, nil
	case IsType:
		args = append(args, vv)
		buf.WriteString(fmt.Sprintf("jsonb_typeof((%s)::jsonb) = LOWER(%s)", name, variable))
		return buf.String(), args, nil
	case IsObject:
		buf.WriteString(fmt.Sprintf("(%s)::jsonb @> '{}'::jsonb", name))
		return buf.String(), args, nil
	case IsArray:
		buf.WriteString(fmt.Sprintf("(%s)::jsonb @> '[]'::jsonb", name))
		return buf.String(), args, nil
	default:
		return "", nil, fmt.Errorf("unsupported operator")
	}

	args = append(args, p.JSONMarshal(vv))
	return buf.String(), args, nil
}

func (p postgres) Value(it interface{}) string {
	var str string
	switch vi := it.(type) {
	case nil:
		str = "NULL"
	case json.RawMessage:
		str = fmt.Sprintf(`'%s'`, escapeSingleQuote(fmt.Sprintf(`%s`, vi)))
	case string, []byte:
		str = fmt.Sprintf(`'%s'`, escapeSingleQuote(fmt.Sprintf(`%s`, vi)))
	default:
		str = fmt.Sprintf("%v", vi)
	}
	return str
}

// DataType :
func (p postgres) DataType(sc Schema) string {
	buf := new(bytes.Buffer)
	buf.WriteString(sc.DataType)
	if sc.IsUnsigned {
		buf.WriteString(fmt.Sprintf(" CHECK (%s >= 0)", p.Quote(sc.Name)))
	}
	if !sc.IsNullable {
		buf.WriteString(" NOT NULL")
		t := reflect.TypeOf(sc.DefaultValue)
		if t != reflect.TypeOf(OmitDefault(nil)) {
			buf.WriteString(fmt.Sprintf(" DEFAULT %s", p.ToString(sc.DefaultValue)))
		}
	}
	return buf.String()
}

func (p postgres) OnConflictUpdate(table string, cols []string) string {
	buf := new(bytes.Buffer)
	buf.WriteString(fmt.Sprintf("ON CONFLICT (%s) DO UPDATE SET ", p.Quote(pkColumn)))
	for _, c := range cols {
		buf.WriteString(fmt.Sprintf("%s = %s.%s,", p.Quote(c), p.GetTable(table), p.Quote(c)))
	}
	buf.Truncate(buf.Len() - 1)
	return buf.String()
}

func (p postgres) GetSchema(c Column) Schema {
	f := c.field
	root := f.getRoot()
	t := root.typeOf
	if root.isFlatten() {
		if !root.isSlice() {
			t = f.typeOf
		}
	}

	sc := Schema{
		Name:       c.Name(),
		IsNullable: f.isPtrChild,
	}

	if t.Kind() == reflect.Ptr {
		sc.IsNullable = true
		if t == typeOfPtrKey {
			sc.IsIndexed = true
			sc.DataType = fmt.Sprintf("varchar(%d)", pkLen)
			sc.CharSet = latin1CharSet
			if f.name == keyFieldName {
				sc.Name = pkColumn
				sc.DefaultValue = OmitDefault(nil)
				sc.IsIndexed = false
			}
			return sc
		}
		t = t.Elem()
	}

	switch t {
	case typeOfJSONRawMessage:
		sc.DefaultValue = OmitDefault(nil)
		sc.DataType = "jsonb"
	case typeOfByte:
		sc.DefaultValue = OmitDefault(nil)
		sc.DataType = "bytea"
	case typeOfDate:
		sc.DefaultValue = "0001-01-01"
		sc.DataType = "date"
	case typeOfTime:
		sc.DefaultValue = time.Time{}
		sc.DataType = "timestamp"
	case typeOfSoftDelete:
		sc.DefaultValue = OmitDefault(nil)
		sc.IsNullable = true
		sc.IsIndexed = true
		sc.DataType = "timestamp"
	default:
		switch t.Kind() {
		case reflect.String:
			sc.DefaultValue = ""
			sc.DataType = fmt.Sprintf("varchar(%d)", 191)
			if f.IsLongText() {
				sc.DefaultValue = nil
				sc.DataType = "text"
			}
		case reflect.Bool:
			sc.DefaultValue = false
			sc.DataType = "bool"
		case reflect.Int:
			sc.DefaultValue = int(0)
			sc.DataType = "integer"
		case reflect.Int8:
			sc.DefaultValue = int8(0)
			sc.DataType = "smallint"
		case reflect.Int16:
			sc.DefaultValue = int16(0)
			sc.DataType = "smallint"
		case reflect.Int32:
			sc.DefaultValue = int32(0)
			sc.DataType = "integer"
		case reflect.Int64:
			sc.DefaultValue = int64(0)
			sc.DataType = "bigint"
		case reflect.Uint:
			sc.DefaultValue = uint(0)
			sc.DataType = "integer"
			sc.IsUnsigned = true
		case reflect.Uint8:
			sc.DefaultValue = uint8(0)
			sc.DataType = "smallint"
			sc.IsUnsigned = true
		case reflect.Uint16:
			sc.DefaultValue = uint16(0)
			sc.DataType = "smallint"
			sc.IsUnsigned = true
		case reflect.Uint32:
			sc.DefaultValue = uint32(0)
			sc.DataType = "integer"
			sc.IsUnsigned = true
		case reflect.Uint64:
			sc.DefaultValue = uint64(0)
			sc.DataType = "bigint"
			sc.IsUnsigned = true
		case reflect.Float32, reflect.Float64:
			sc.DefaultValue = float64(0)
			sc.DataType = "real"
		default:
			sc.DataType = "jsonb"
		}
	}

	return sc
}

// GetColumns :
func (p *postgres) GetColumns(table string) (columns []string) {
	stmt := "SELECT column_name FROM INFORMATION_SCHEMA.columns WHERE table_schema = CURRENT_SCHEMA() AND table_name = $1;"
	rows, _ := p.db.Query(stmt, table)
	defer rows.Close()
	for i := 0; rows.Next(); i++ {
		columns = append(columns, "")
		rows.Scan(&columns[i])
	}
	return
}

// GetIndexes :
func (p *postgres) GetIndexes(table string) (idxs []string) {
	stmt := "SELECT indexname FROM pg_indexes WHERE schemaname = CURRENT_SCHEMA() AND tablename = $1;"
	rows, _ := p.db.Query(stmt, table)
	defer rows.Close()
	for i := 0; rows.Next(); i++ {
		idxs = append(idxs, "")
		rows.Scan(&idxs[i])
	}
	return
}

func (p *postgres) HasTable(table string) bool {
	var count int
	p.db.QueryRow("SELECT count(*) FROM INFORMATION_SCHEMA.tables WHERE table_type = 'BASE TABLE' AND table_schema = CURRENT_SCHEMA() AND table_name = $1;", table).Scan(&count)
	return count > 0
}

func (p *postgres) HasIndex(table, idx string) bool {
	var count int
	p.db.QueryRow("SELECT count(*) FROM pg_indexes WHERE tablename = $1 AND indexname = $2 AND schemaname = CURRENT_SCHEMA()", table, idx).Scan(&count)
	return count > 0
}

func (p *postgres) ToString(it interface{}) string {
	var v string
	switch vi := it.(type) {
	case nil:
		v = "NULL"
	case json.RawMessage:
		v = fmt.Sprintf(`'%s'`, vi)
	case string:
		v = fmt.Sprintf(`'%s'`, vi)
	case bool:
		v = fmt.Sprintf("%t", vi)
	case uint, uint8, uint16, uint32, uint64:
		v = fmt.Sprintf("%d", vi)
	case int, int8, int16, int32, int64:
		v = fmt.Sprintf("%d", vi)
	case float32, float64:
		v = fmt.Sprintf("%v", vi)
	case time.Time:
		v = fmt.Sprintf(`'%s'`, vi.Format("2006-01-02 15:04:05"))
	case []interface{}:
		v = fmt.Sprintf(`'%s'`, "[]")
	default:
		v = fmt.Sprintf("%v", vi)
	}
	return v
}

func (p *postgres) CreateTable(table string, columns []Column) error {
	idxs := make([]string, 0, len(columns))
	conn := p.db.sqlCommon.(*sql.DB)
	tx, err := conn.Begin()
	if err != nil {
		return err
	}
	defer tx.Rollback()

	buf := new(bytes.Buffer)
	buf.WriteString(fmt.Sprintf("CREATE TABLE IF NOT EXISTS %s (", p.GetTable(table)))
	for _, c := range columns {
		ss := p.GetSchema(c)
		buf.WriteString(fmt.Sprintf("%s %s,",
			p.Quote(ss.Name),
			p.DataType(ss)))

		if ss.IsIndexed {
			idx := fmt.Sprintf("%s_%s_%s", table, ss.Name, "Idx")
			stmt := fmt.Sprintf("CREATE INDEX %s ON %s (%s);",
				p.Quote(idx), p.GetTable(table), p.Quote(ss.Name))
			idxs = append(idxs, stmt)
		}
	}
	buf.WriteString(fmt.Sprintf("PRIMARY KEY (%s)", p.Quote(pkColumn)))
	buf.WriteString(");")
	log.Println(buf.String())
	if _, err := tx.Exec(buf.String()); err != nil {
		return err
	}

	for _, idx := range idxs {
		if _, err := tx.Exec(idx); err != nil {
			return err
		}
	}

	return tx.Commit()
}

func (p *postgres) AlterTable(table string, columns []Column) error {
	cols := newDictionary(p.GetColumns(table))
	idxs := newDictionary(p.GetIndexes(table))
	idxs.delete(fmt.Sprintf("%s_pkey", table))
	buf := new(bytes.Buffer)
	buf.WriteString(fmt.Sprintf("ALTER TABLE %s ", p.GetTable(table)))
	for _, c := range columns {
		ss := p.GetSchema(c)
		if !cols.has(ss.Name) {
			buf.WriteString(fmt.Sprintf("ADD COLUMN %s %s", p.Quote(ss.Name), ss.DataType))
			if !ss.IsNullable {
				buf.WriteString(" NOT NULL")
				if !ss.IsOmitEmpty() {
					buf.WriteString(fmt.Sprintf(" DEFAULT %s",
						p.ToString(ss.DefaultValue)))
				}
			}
			buf.WriteString(",")
		} else {
			prefix := fmt.Sprintf("ALTER COLUMN %s", p.Quote(ss.Name))
			buf.WriteString(fmt.Sprintf("%s TYPE %s", prefix, ss.DataType))
			buf.WriteString(",")
			if !ss.IsNullable {
				buf.WriteString(prefix + " SET NOT NULL,")
				if !ss.IsOmitEmpty() {
					buf.WriteString(fmt.Sprintf("%s SET DEFAULT %s,",
						prefix, p.ToString(ss.DefaultValue)))
				}
			}
		}

		if ss.IsIndexed {
			idx := fmt.Sprintf("%s_%s_%s", table, ss.Name, "idx")
			if idxs.has(idx) {
				idxs.delete(idx)
			} else {

				// buf.WriteString(fmt.Sprintf(
				// 	" CREATE INDEX %s ON (%s);",
				// 	p.Quote(idx),
				// 	p.Quote(ss.Name)))
			}
		}
		cols.delete(ss.Name)

	}

	for _, col := range cols.keys() {
		buf.WriteString(fmt.Sprintf(" DROP COLUMN %s,", p.Quote(col)))
	}

	buf.Truncate(buf.Len() - 1)
	buf.WriteString(";")

	return p.db.ExecStmt(&Stmt{
		query: buf,
	})

	// for _, idx := range idxs.keys() {
	// 	buff := new(bytes.Buffer)
	// 	buff.WriteString(fmt.Sprintf("DROP INDEX %s;", p.Quote(idx)))
	// 	p.db.ConsoleLog(&Stmt{buff, nil, nil})
	// 	if _, err := tx.Exec(buff.String()); err != nil {
	// 		return err
	// 	}
	// }

	// for _, idx := range idxs.keys() {
	// 	stmt := fmt.Sprintf("CREATE INDEX %s ON %s ();", p.Quote(idx), p.Quote(table))
	// 	if _, err := tx.Exec(stmt); err != nil {
	// 		return err
	// 	}
	// }
}

func (p *postgres) ReplaceInto(src, dst string) error {
	cols := p.GetColumns(src)
	pk := p.Quote(pkColumn)
	src, dst = p.GetTable(src), p.GetTable(dst)

	buf := new(bytes.Buffer)
	buf.WriteString("WITH patch AS (")
	buf.WriteString("UPDATE ")
	buf.WriteString(dst)
	buf.WriteString(" SET ")

	for _, c := range cols {
		if c == pkColumn {
			continue
		}
		cc := p.Quote(c)
		buf.WriteString(cc)
		buf.WriteString(" = ")
		buf.WriteString(src + "." + cc)
		buf.WriteString(",")
	}

	buf.Truncate(buf.Len() - 1)
	buf.WriteString(" FROM ")
	buf.WriteString(src)
	buf.WriteString(" WHERE " + src + "." + pk + " = " + dst + "." + pk + " ")
	buf.WriteString("RETURNING " + src + "." + pk + ") ")
	buf.WriteString("INSERT INTO ")
	buf.WriteString(dst)
	buf.WriteString(" SELECT * FROM ")
	buf.WriteString(src)
	buf.WriteString(" WHERE NOT EXISTS ")
	buf.WriteString("(SELECT 1 FROM patch WHERE " + pk + " = " + src + "." + pk + ")")
	buf.WriteString(";")

	return p.db.ExecStmt(&Stmt{
		query: buf,
	})
}
